### DISCUSSION TITLE
Integration Event vs Notification Event

### DISCUSSION THREAD

**Discussion by mehdihadeli** [kgrzybek/modular-monolith-with-ddd#168] [Q&A] [answered]
Hi @kgrzybek,
What is the difference between `Integration Event` and `Notification Event` in your code base?

**Comment by AndreiGanichev** [comment, marked as answer]
Hello, @mehdihadeli.
_Domain Event Notification_ is based on the DDD tactical pattern _Domain event_ and is a way to handle a domain event outside the transaction which it was published in. So Domain Event Notification "lives" inside _Bounded context_(BC). _Integration Event_ on the contrary is a part of _Published language_ of BC and is a way for different BCs to communicate with each other.

As for differences in code, Domain events are part of Domain layer, Domain Event Notification - Application layer and each module declares its Integration Events in separate project `Xxx.IntegrationEvents`. Moreover Domain events(and so Domain Event Notification) can contain _Value objects_(typed Id for example) but Integration Events consists of primitive types to be simple serializible and not to reveal BC's internals outside.
I should mention that sometimes it's reasonable even for Domain events to consist of only primitive types. Event soursing imply to store all events and allow to recover Aggregate state by replaying the sequence of events. So it's crucial to be able to deserialize any event from the past. Using Value objects can cause incorrect state if some logic has been changed since an event happend.

For more details I advice to read @kgrzybek's blog [post](http://www.kamilgrzybek.com/design/handling-domain-events-missing-part/)

**Reply by mehdihadeli** [reply]
@AndreiGanichev Thanks for your explanation, 
Could you share your resource for `DDD tactical pattern Domain event`?
But why Notification events placed inside application layer?

**Reply by AndreiGanichev** [reply]
1. As for theory about Domain events I would recommend _Chapter 8. Domain Events_ in _Implementing Domain Driven Design by Vaughn Vernon_ aka IDDD. To understand ways of handling Domain event, transactional and eventual consistency you can also read _Chapter 10. Aggregates_.

2. On one hand Domain Event Notification is based on Domain event but on another it's more a mechanism then smth domain experts care about(unlike Domain events themself). It's a mechanism to specify logic which handles Domain event in separate transaction. And it's Application layer concern to manage transaction boandaries:
> Remember, the Application Service controls the transaction. (IDDD. Chapter 8. Domain Events)

**Reply by mehdihadeli** [reply]
@AndreiGanichev Thanks a lot.
